gcc/rust/checks/lints/unused-var/rust-unused-var-collector.cc

```diff
@@ -0,0 +1,130 @@
+// Copyright (C) 2025 Free Software Foundation, Inc.
+
+// This file is part of GCC.
+
+// GCC is free software; you can redistribute it and/or modify it under
+// the terms of the GNU General Public License as published by the Free
+// Software Foundation; either version 3, or (at your option) any later
+// version.
+
+// GCC is distributed in the hope that it will be useful, but WITHOUT ANY
+// WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+// for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with GCC; see the file COPYING3.  If not see
+// <http://www.gnu.org/licenses/>.
+
+#include "rust-unused-var-collector.h"
+#include "rust-hir-full-decls.h"
+#include "rust-hir-item.h"
+#include "rust-hir-path.h"
+#include "rust-immutable-name-resolution-context.h"
+
+namespace Rust {
+namespace Analysis {
+UnusedVarCollector::UnusedVarCollector ()
+  : nr_context (
+    Resolver2_0::ImmutableNameResolutionContext::get ().resolver ()),
+    mappings (Analysis::Mappings::get ()),
+    unused_var_context (*UnusedVarContext::get ())
+{}
+void
+UnusedVarCollector::go (HIR::Crate &crate)
+{
+  for (auto &item : crate.get_items ())
+    item->accept_vis (*this);
+}
+void
+UnusedVarCollector::visit (HIR::LetStmt &stmt)
+{
+  HIR::Pattern &pattern = stmt.get_pattern ();
+  collect_variable (pattern);
```

Comment:
can't we visit instead here? mabye I'm missing something but I think all patterns should create variables to check for, so we could just do `visit(pattern)` and then create visitors for all our kinds of patterns instead of manually dispatching like in `collect_variable`